src/jrd/btr.cpp

```diff
@@ -2433,78 +2455,164 @@ static void compress(thread_db* tdbb,
 		if (descending)
 		{
 			// DESC NULLs are stored as 1 byte
-			*p++ = pad;
-			key->key_length = (p - key->key_data);
+			key->key_data[0] = pad;
+			key->key_length = 1;
 		}
 		else
 			key->key_length = 0; // ASC NULLs are stored with no data
 
+		fb_assert(!key->key_next);
 		return;
 	}
 
+	// For descending index and new index structure we insert 0xFE at the beginning.
+	// This is only done for values which begin with 0xFE (254) or 0xFF (255) and
+	// is needed to make a difference between a NULL state and a VALUE.
+	// Note! By descending index key is complemented after this compression routine.
+	// Further a NULL state is always returned as 1 byte 0xFF (descending index).
+	const UCHAR desc_end_value_prefix = 0x01; // ~0xFE
+	const UCHAR desc_end_value_check = 0x00; // ~0xFF;
+
+	const Database* dbb = tdbb->getDatabase();
+	bool first_key = true;
+	VaryStr<MAX_KEY * 4> buffer;
+	size_t multiKeyLength;
+	UCHAR* ptr;
+	UCHAR* p = key->key_data;
+
 	if (itype == idx_string || itype == idx_byte_array || itype == idx_metadata ||
 		itype >= idx_first_intl_string)
 	{
-		VaryStr<MAX_KEY> buffer;
-		const UCHAR pad = (itype == idx_string) ? ' ' : 0;
-		UCHAR* ptr;
+		temporary_key* root_key = key;
+		bool has_next;
 
-		size_t length;
-
-		if (itype >= idx_first_intl_string || itype == idx_metadata)
+		do
 		{
-			DSC to;
-
-			// convert to an international byte array
-			to.dsc_dtype = dtype_text;
-			to.dsc_flags = 0;
-			to.dsc_sub_type = 0;
-			to.dsc_scale = 0;
-			to.dsc_ttype() = ttype_sort_key;
-			to.dsc_length = MIN(MAX_KEY, sizeof(buffer));
-			ptr = to.dsc_address = reinterpret_cast<UCHAR*>(buffer.vary_string);
-			length = INTL_string_to_key(tdbb, itype, desc, &to, key_type);
-		}
-		else
-			length = MOV_get_string(desc, &ptr, &buffer, MAX_KEY);
+			size_t length;
 
-		if (length)
-		{
-			// clear key_empty flag, because length is >= 1
-			key->key_flags &= ~key_empty;
+			has_next = false;
 
-			if (length > sizeof(key->key_data))
-				length = sizeof(key->key_data);
+			if (first_key)
+			{
+				first_key = false;
 
-			if (descending && ((*ptr == desc_end_value_prefix) || (*ptr == desc_end_value_check)))
+				if (itype >= idx_first_intl_string || itype == idx_metadata)
+				{
+					DSC to;
+
+					// convert to an international byte array
+					to.dsc_dtype = dtype_text;
+					to.dsc_flags = 0;
+					to.dsc_sub_type = 0;
+					to.dsc_scale = 0;
+					to.dsc_ttype() = ttype_sort_key;
+					to.dsc_length = MAX_KEY * 4;
```

Comment:
In FB4, `MAX_KEY * 4 == 32768` while `MAX_COLUMN_SIZE == 32767`. Are you sure `dsc_length > MAX_COLUMN_SIZE` is safe? At least `MOV_make_string2` could throw as a result.